libs/domain/push-notification/src/services/providers/web-push/web-push.spec.ts

```diff
@@ -0,0 +1,202 @@
+import { createInjector, destroyInjector } from '@spryker-oryx/di';
+import { of } from 'rxjs';
+import { PushProvider } from '../';
+import { WebPushProvider } from './web-push';
+
+class PushManagerMock
+  implements Pick<PushManager, 'subscribe' | 'getSubscription'>
+{
+  subscription = new PushSubscriptionMock();
+  subscribe = vi.fn().mockReturnValue(of(this.subscription));
+  getSubscription = vi.fn();
+}
+
+class PushSubscriptionMock
+  implements Pick<PushSubscription, 'unsubscribe' | 'toJSON'>
+{
+  toJSON = vi.fn();
+  unsubscribe = vi.fn();
+}
+
+const mockServiceWorker = {
+  ready: new Promise<any>((resolve) =>
+    resolve({ pushManager: new PushManagerMock() })
+  ),
+};
+
+Object.defineProperty(navigator, 'serviceWorker', {
+  value: mockServiceWorker,
+  configurable: true,
+});
+
+describe('WebPushProvider', () => {
+  async function getPushManager() {
+    const serviceWorker = await navigator.serviceWorker.ready;
+    return serviceWorker.pushManager as unknown as PushManagerMock;
+  }
+
+  let provider: WebPushProvider;
+
+  beforeEach(() => {
+    const testInjector = createInjector({
+      providers: [
+        {
+          provide: PushProvider,
+          useClass: WebPushProvider,
+        },
+      ],
+    });
+
+    provider = testInjector.inject(PushProvider) as WebPushProvider;
+  });
+
+  afterEach(() => {
+    vi.clearAllMocks();
+    destroyInjector();
+  });
+
+  it('should be provided', () => {
+    expect(provider).toBeInstanceOf(WebPushProvider);
+  });
+
+  describe('init() method', () => {
+    it('should immediately resolve', () => {
+      const callback = vi.fn();
+
+      provider.init().subscribe(callback);
+
+      expect(callback).toHaveBeenCalled();
+    });
+  });
+
+  describe('getSubscription() method', () => {
+    let pushManager: PushManagerMock;
+    beforeEach(async () => {
+      pushManager = await getPushManager();
+    });
+
+    it('should create and return Promise of subscription data from PushManager', async () => {
+      const mockSubscription = { endpoint: 'mock-endpoint' };
+
+      pushManager.subscription.toJSON.mockReturnValue(mockSubscription);
+      pushManager.getSubscription.mockReturnValue(of(undefined));
+
+      provider.getSubscription().subscribe((subscription) => {
+        expect(pushManager.subscribe).toHaveBeenCalledWith({
+          userVisibleOnly: true,
+        });
+        expect(subscription).toEqual(mockSubscription);
+        // Should be a different object then original subscription
+        expect(subscription).not.toEqual(pushManager.subscription);
+      });
```

Comment:
Please do not use `expect` inside of the callbacks like `.subscribe()` because if the callback is not invoked none of the expectations will be ever triggered.
You can use a spy and then check that it was called with specific arguments using `.toHabeenCalledWith()` API.

Reply:
I think the simplest solution is to wait for next tick in the test or if it's promises only then `await Promise.resolve()` should be even better and slightly faster as you are going to execute the test in the single tick still.